compiler/resolution/nilChecking.cpp

```diff
@@ -644,6 +724,111 @@ static Expr* getSingleDefExpr(Expr* expr) {
   return NULL;
 }
 
+// SE->symbol() is known to be nil/non-nil. If it comes from a MOVE or CAST,
+// propagate to their arguments.
+// For efficiency, the caller should assure isSymbolAnalyzed(SE).
+// Returns true and skips updating 'OUT' if the branch in question is unreachable.
+static bool adjustTestArgChain(SymExpr* SE, bool isAllocated, AliasMap& OUT) {
+  SymExpr* curr = SE;
+  while (true) {
+    AliasMap::const_iterator it = OUT.find(curr->symbol());
+    if (it != OUT.end()) {
+      AliasType altype = it->second.type;
+      if ((isAllocated && altype == MUST_ALIAS_NIL)        ||
+          (!isAllocated && altype == MUST_ALIAS_ALLOCATED) )
+        // We are testing ==nil on something that's not nil, or visa versa.
+        return true;
+    }
+
+    // Not worth updating OUT if 'curr' is the only use (a common case?).
+    if (curr->symbol()->getSingleUse() == curr) ;
```

Comment:
Wouldn't it be better to write it as `if (curr->symbol()->getSingleUse() != curr) updateForConditional(...)` ?

Does this actually matter for performance? For correctness?